TestResultSummaryService/parsers/BenchmarkParser.js

```diff
@@ -2,169 +2,135 @@ const Parser = require( './Parser' );
 const BenchmarkMetricRouter = require( './BenchmarkMetricRouter' );
 const BenchmarkMetric = require( './BenchmarkMetric' );
 
-const benchmarkDelimiterRegex = /\*\*\*\*\*\*\*\*\*\* START OF NEW TESTCI BENCHMARK JOB \*\*\*\*\*\*\*\*\*\*[\r\n]/;
-const benchmarkNameRegex = /Benchmark Name: (.*) Benchmark Variant: .*[\r\n]/;
-const benchmarkVariantRegex = /Benchmark Name: .* Benchmark Variant: (.*)[\r\n]/;
-const productResourceRegex = /Product Resource: (.*)[\r\n]/;
-
 class BenchmarkParser extends Parser {
 
-    // True iff all benchmarks and their corresponding iterations in the given run can be parsed
-    static canParse( buildName, output ) {
-        return benchmarkDelimiterRegex.test(output);
+    static getBenchmarkInfo( testName ) {
+        let result, benchmarkName, benchmarkVariant;
+
+        /* For PerfNext, benchmark name is allowed to have multiple underscores but variant shouldn't have underscores
+         * in order for TRSS to parse the job. For TKG, we'll be good since we don't specify variant name anyway.
+         */
+        if (testName) {
+            if ((result = testName.split("_")) != null) {
+                if (Array.isArray(result) && result.length > 0) {
+                    benchmarkVariant = result[result.length - 1];
+                    benchmarkName = result[0];
+                    let i = 1;
+                    for (; i < result.length - 1; i++) {
+                        benchmarkName = benchmarkName + "_" + result[i];
+                    }
+                }
+            }
+        }
+        return {benchmarkName, benchmarkVariant};
     }
 
-    // Returns an iterator object which accesses each benchmark iteration from the given output
-    static getBenchmarkIterator(output) {
-        let index = 1;
-        let splitOutput = output.split(benchmarkDelimiterRegex);
-        let splitOutputLength = splitOutput.length;
-    
-        // No delimiter inside the output
-        if (!(Array.isArray(splitOutput)) || splitOutputLength === 1) {
-            return null;
+    // True iff all benchmarks and their corresponding iterations in the given run can be parsed
+    static canParse( testName ) {
```

Comment:
This `canParse ()` function is used to check all test names now, not only for perf tests. Please make it efficient. 

For example,
- I do not understand the purpose of `getBenchmarkInfo()`. Why do we need to split the `testName ` by `_`, loop through it and add it back with `_`?  If we just need to separate last `_`, can we use `lastIndexOf("_")`?
- there is no need to check `benchmarkName` and `benchmarkVariant` in every if block.
- if possible, I think we should remove `benchmarkVariant ` completely, so BenchmarkMetricRouter.js and other related code can be much simpler.

Reply:
I just double checked files`BenchmarkMetriRouter.js` and `BenchmarkMetric.js`. Different benchmarkName/benchmarkVariants may direct to have the same metrics. Since we wanna run as many benchmarks(I believe so) as possible, I think it might not be possible to remove benchmarkVariants all here at a time.